docs/enhancements/dynamic-fetching-of-openshift-versions.md

```diff
@@ -0,0 +1,280 @@
+* As described above, we do version updates either automatically or manually.
+* QE team wants to check if a newly added OpenShift version has a regression in its
+  behavior from a previous version.
+* QE team tests a flow which includes OpenShift upgrade process.
+
+### Goals
+
+* Assisted-service is able to provide all versions included in the upgrade-channels,
+  based on provided configuration.
+* We keep ourselves the option to specifically configure certain versions as we can
+  nowadays. This is to make sure we can handle edge-cases and be flexible if needed.
+* UI needs to be adjusted to handle this growing list of versions (at the time being,
+  a total of ~140 versions).
+
+### Non-Goals
+
+* Dynamic handling of y-versions, live-iso discovery images, and such. A new minor
+  OpenShift version is introduced every 4 months, requires additional code changes
+  (for example in feature-support-level list) and it isn't worth the costs and
+  implications of making assisted-image-service refresh itself and pull-in new
+  images.
+* Handling OpenShift versions in a kubernetes API interface. For this use-case
+  versions are externally configured in the ``ClusterImageSet`` custom resource.
+
+## Proposal
+
+The assisted-service will make use of a new configuration variable called
+``DYNAMIC_OPENSHIFT_RELEASES_CONFIG``, which outlines the supported matrix. For
+example:
+
+```
+DYNAMIC_OPENSHIFT_RELEASES_CONFIG='{
+   "4.11": {
+      "upgrade_channels": [
+         {
+            "cpu_architecture": "x86_64",
+            "channels": ["stable"]
+         }
+      ]
+   },
+   "4.12": {
+      "upgrade_channels": [
+         {
+            "cpu_architecture": "x86_64",
+            "channels": ["candidate", "stable"]
+         },
+         {
+            "cpu_architecture": "arm64",
+            "channels": ["candidate", "stable"]
+         },
+         {
+            "cpu_architecture": "multi",
+            "channels": ["candidate", "stable"]
+         }
+      ]
+   }
+}'
+```
+
+The assisted-service periodically polls versions from the ``upgrades_info`` API
+in the form of ``https://api.openshift.com/api/upgrades_info/v1/graph?channel=<channel>&arch=<arch>``.
+For example:
+```
+$ curl -s 'https://api.openshift.com/api/upgrades_info/v1/graph?channel=stable-4.11&arch=amd64' \
+    | jq -r '.nodes[] | select(.version | startswith("4.11")) | .version'
+4.11.9
+4.11.4
+4.11.8
+...
+```
+
+It then adds any new version found by this process in the database. When a
+consumer of the API requests the available versions via ``/v2/openshift-versions``
+(either it's the UI or a direct REST request) versions are read from the
+database.
+
+To help fetching only certain versions via auto-completion like U/X, we should
+provide a parameter ``version_pattern`` for the ``/v2/openshift-versions``
+endpoint. For example:
+```
+/v2/openshift-versions:
+   get:
+      ...
+      parameters:
+        - in: query
+          name: version_pattern
```

Comment:
Should we support multiple 'version_pattern'? E.g. So the UI could fetch all vers in a single request.

Reply:
they can do it by not providing the ``version_pattern`` parameter (or having it empty)